### ISSUE TITLE
Question about the power iteration algorithm.

### ISSUE THREAD

**Issue by XuyangBai** [Thinklab-SJTU/PCA-GM#18]
Hi @rogerwwww Thanks for your sharing. I am not from the graph matching area, so sorry if my question is stupid. I know that the power iteration algorithm is to efficiently calculate the leading eigenvector of the matrix `M`, but when I try to use the `torch.symeig()` and find the largest eigenvector, I find that sometimes the power iteration algorithm can not converge to the value found by `torch.symeig()` no matter how many iterations. Another problem is the gradient of power iteration and `torch.symeig()` are quite different. So I am really confused about the usage of the power iteration algorithm, what's the main reason for using power iteration to find the leading eigenvector instead of `torch.symeig()`?

You may try the code here.
```python
import numpy as np
import torch
import time

np.set_printoptions(precision=3)
np.random.seed(123)

def check_eigen(num_dim, max_iter):
    # random matrix (positive)
    matrix_np = np.abs(np.random.randn(num_dim,num_dim))
    # make symmetric
    matrix_np = matrix_np + matrix_np.T
    matrix_torch = torch.autograd.Variable(torch.from_numpy(matrix_np), requires_grad=True)
    matrix_pi = torch.autograd.Variable(torch.from_numpy(matrix_np), requires_grad=True)

    
    # compute eigenvalue decompositions using numpy
    eigvals_np, eigvecs_np = np.linalg.eigh(matrix_np)
    # compute eigenvalue decompositions using PyTorch
    stime = time.time()
    eigvals_torch, eigvecs_torch = torch.symeig(matrix_torch, eigenvectors=True, upper=True)
    etime = time.time()
    time_eig = etime - stime
    leading_eig_torch = eigvecs_torch[:,-1]
    # print(f"Torch Leading EigenVector {etime-stime:.10f}s:", eigvecs_torch[:,-1].data.numpy())
    assert np.allclose(eigvals_np, eigvals_torch.data.numpy())

    # Power Iteration Algorithm 
    stime = time.time()
    solution = torch.ones_like(matrix_pi[:, 0:1])
    stop_iteration = max_iter
    for i in range(max_iter):
        solution = matrix_pi @ solution
        solution = solution / torch.norm(solution)
        error = float(torch.norm(leading_eig_torch - solution.squeeze(), p=2)) 
        if error < 1e-6:
            stop_iteration = i
            etime = time.time()
            time_pi = etime - stime
            print(f"Converge at iter-{i} with error={error:.10f}")
            break
    if i == max_iter-1:
        etime = time.time()
        time_pi = etime - stime
    solution = solution.squeeze(-1)

    # comparing the gradient of Power iteration and torch.symeig
    leading_eig_torch[0].backward()
    grad_eig = matrix_torch.grad.numpy()
    solution[0].backward()
    grad_pi = matrix_pi.grad.numpy()
    # print(np.linalg.norm(grad_eig - grad_pi))
    # assert np.allclose(grad_eig, grad_pi)
    return time_eig, time_pi, stop_iteration, float(i!=max_iter-1)
    

if __name__ == '__main__':
    time_eig_list = []
    time_pi_list = []
    stop_iteration_list = []
    converegence_list = []
    for i in range(200, 400):
        time_eig, time_pi, stop_iteration, flag = check_eigen(i, max_iter=500)
        time_eig_list.append(time_eig)
        time_pi_list.append(time_pi)
        stop_iteration_list.append(stop_iteration)
        converegence_list.append(flag)
    print(f"Eig mean time: {np.mean(time_eig_list):.6f}s")
    print(f"PI mean time: {np.mean(time_pi_list):.6f}s")
    print(f"PI mean iter: {np.mean(stop_iteration_list):.6f}")
    print(f"PI convergance ratio: {np.mean(converegence_list):.2f}")
``` 
Thanks in advance.

**Comment by rogerwwww** [comment]
For the first question about the convergence of power iteration, it is because the leading eigenvalue can be either positive or negative. Therefore, some modifications should be made when computing the leading eigenvector for torch: 
```python
leading_eig_torch = eigvecs_torch[:, torch.argmax(torch.abs(eigvals_torch))]
```
And for the same reason, the power iteration algorithm may converge to either ``leading_eig_torch`` or ``-leading_eig_torch`` and both of them should be considered. I also adopt the common practice when determining the convergence (the solution does not change much). Therefore, I modified your power iteration algorithm as:
```python
    for i in range(max_iter):
        last_solution = solution
        solution = matrix_pi @ solution
        solution = solution / torch.norm(solution)
        error = float(torch.norm(leading_eig_torch - solution.squeeze(), p=2))
        error2 = float(torch.norm(leading_eig_torch + solution.squeeze(), p=2))
        if torch.norm(solution - last_solution) < 1e-10:
            stop_iteration = i
            etime = time.time()
            time_pi = etime - stime
            print("Converge at iter-{} with error={:.10f}, error2={:.10f}".format(i, error, error2))
            break
```
and either ``error`` or ``error2`` equals to 0 when running your code. 

For the second question about the backward gradient, I also discover such a phenomenon but unfortunately I fail to come up with any possible reasons. Any discussion is welcomed if you have any ideas!

**Comment by XuyangBai** [comment]
Hi @rogerwwww Thanks a lot for pointing out the sign of leading eigenvector. I will reopen this issue if I have some updates on the second question.
